src/typechecker.erl

```diff
@@ -834,9 +841,9 @@ expand_builtin_aliases({type, Ann, maybe_improper_list, []}) ->
 expand_builtin_aliases({type, Ann, nonempty_list, []}) ->
     {type, Ann, nonempty_list, [{type, Ann, any, []}]};
 expand_builtin_aliases({type, Ann, string, []}) ->
-    {type, Ann, list, [{type, Ann, char, []}]};
+    {type, Ann, list, [expand_builtin_aliases({type, Ann, char, []})]};
 expand_builtin_aliases({type, Ann, nonempty_string, []}) ->
-    {type, Ann, nonempty_list, [{type, Ann, char, []}]};
+    {type, Ann, nonempty_list, [expand_builtin_aliases({type, Ann, char, []})]};
```

Comment:
I don't mind this, but I don't think it's necessary if it's only used by normalize. It shouldn't look inside the element type.

Reply:
This changed was prompted by our discussion in the other issue - https://github.com/josefs/Gradualizer/pull/369#issuecomment-1024977834. Specifically:

>> Do you have an example in which this turns into infinite recursion?

> I think the problem here is not infinite recursion - ultimately builtin type expansion is well defined and ends at some point. However, I remember seeing a buggy check result, which when traced was caused by a mismatch between a string() and a [char()] - or was it [char()] vs [0..16#10ffff]? - the point is it was a mismatch between a more and less expanded builtin type. I'll find an exact example.

We have to expand "to the bottom", as otherwise we'll get half-expanded builtin types which don't compare equal, where they should, and therefore false positives in the output.

Reply:
BTW, check the commit message - this change is introduced by "Fix exhaustive_refinable_map_variants":

```
$ fd exhaustive_refinable_map_variants
test/should_fail/exhaustive_refinable_map_variants.erl
```